### PR TITLE
Bulk shifting

### PR DESCRIPTION
**Pull request by madamczyk-intel**
* shifting is implemented by roll'ing by a predefined list of chunks
* kv-cache is shifted in bulk
* right_padding is now calculated based on input_length and seq_length

### CODE REVIEW

**Comment by madamczyk-intel** [comment]
@mswiniarsk please review

File: server/text_generation_server/models/causal_lm.py [hunk 1]
```diff
@@ -96,24 +97,32 @@ def move_data(dst_tensor, chunk_size, indices, src_tensors):
     return result
 
 
+def generate_shift_chunks(offset):
+    chunk_sizes = [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048]
```
Comment (mswiniarsk) [inline server/text_generation_server/models/causal_lm.py:101]: Can you do some list comprehension here so it doesn't look like hardcoded list? 
I guess the max (2048) is based on max total tokens right? So we could use such var and based on that prepare a list

File: server/text_generation_server/models/causal_lm.py [hunk 2]
```diff
@@ -265,52 +283,59 @@ def to_tensors(ind): return (torch.tensor(ind[0], device=device), torch.tensor(i
...
+        src_keys = shift_all(src_keys, key_dim, offsets)
+        htorch.core.mark_step()
+        src_keys = [[t.squeeze(0).clone() for t in torch.split(src, 1)] for src in src_keys]
+        htorch.core.mark_step()
 
-            past_key_values.append((updated_key, updated_value))
-            for b in batches:
-                b.past_key_values[layer_num] = None
+        dst_keys = [prepare_memory(new_bs * chunk_size, prev, inplace) for prev in src_keys[target_batch_idx]]
+        dst_keys = [move_data(dst_keys[layer_num], chunk_size, indices, [src[layer_num] for src in src_keys]) for layer_num in range(num_layers)]
```
Comment (mswiniarsk) [inline server/text_generation_server/models/causal_lm.py:322]: is it more performant to loop over src_keys and src_values multiple times and place mark steps? instead to what we had (single loop over layers)

**Review by mswiniarsk — CHANGES_REQUESTED** [review]
Can you please take a look at my comments?

File: server/text_generation_server/models/causal_lm.py (hunk 1 above)
Comment (madamczyk-intel) [reply server/text_generation_server/models/causal_lm.py:101]: We could do something like: `[2 ** power for power in range(12)]` . Is that ok?
The reason I kept it as a hardcoded list is that the algorithm can work with other values as well (not only powers of 2). In theory if an offset appears often enough and it triggers a long chain we could add it to the list to limit the number of shifts needed on average.

2048 is only loosely based on max_tokens. Even if we had to shift by 4096 we would simply do two shifts by 2048. It's a trade-off between number of recipes and average number of shifts needed.

File: server/text_generation_server/models/causal_lm.py (hunk 2 above)
Comment (madamczyk-intel) [reply server/text_generation_server/models/causal_lm.py:322]: Unfortunately those mark_steps are needed to limit the number of recipes.
In previous approach we called shift 80*2 times and op accumulation + graph starting overhead was dominating run time.

Comment (madamczyk-intel) [reply server/text_generation_server/models/causal_lm.py:322]: I'll double check... Both pad and shift are calling mark_step on their own so in theory we could get rid of some of them...

Comment (madamczyk-intel) [reply server/text_generation_server/models/causal_lm.py:322]: Yeah, some of those mark_steps were redundant. I've removed them.

File: server/text_generation_server/models/causal_lm.py (hunk 1 above)
Comment (mswiniarsk) [reply server/text_generation_server/models/causal_lm.py:101]: Hmm, if it is only loosely connected to max_tokens, then `12` is still a magical number, and I guess both solution are good enough.

**Comment by mswiniarsk** [comment]
LGTM!

**Review by mswiniarsk — APPROVED** [review]

**Pull request merged**